analytics/pages/page_public_dataset.py

```diff
@@ -0,0 +1,258 @@
+                users = selected_users
+            else:
+                users = st.session_state.all_users
+                add_regressions = False
+
+            fig = go.Figure()
+            for u, user in enumerate(users):
+                df_user = df_filt[df_filt["public_username"] == user]
+                fig.add_trace(
+                    go.Scatter(
+                        x=df_user[x],
+                        y=df_user[y],
+                        name=user,
+                        mode="markers",
+                        marker_color=TCOLOR[u % len(TCOLOR)],
+                    )
+                )
+
+                # TODO: Cloud be improved/simplified
+                if add_regressions:
+
+                    df_user.dropna(inplace=True)
+
+                    x_fit = df_user[x].to_numpy().reshape((-1, 1))
+                    y_fit = df_user[y].to_numpy().reshape((-1, 1))
+
+                    model = LinearRegression()
+                    model.fit(x_fit, y_fit)
+
+                    x_range = np.linspace(-10, 10, 100).reshape((-1, 1)).tolist()
+                    y_range = model.predict(x_range).tolist()
+
+                    x_range = [val[0] for val in x_range]
+                    y_range = [val[0] for val in y_range]
+
+                    fig.add_trace(
+                        go.Scatter(
+                            x=x_range,
+                            y=y_range,
+                            name=f"linear reg",
+                            mode="lines",
+                            marker_color=TCOLOR[u % len(TCOLOR)],
+                        )
+                    )
+
+            fig.update_layout(xaxis=dict(title=x), yaxis=dict(title=y))
+            fig.update_xaxes(showline=True, linewidth=2, linecolor="black")
+            fig.update_yaxes(showline=True, linewidth=2, linecolor="black")
+
+            st.plotly_chart(fig)
+
+        else:
+            st.warning(MSG_NO_DATA)
+
+
+def add_expander_cursor_position():
+
+    with st.expander("Cursor position histogram"):
+
+        if isinstance(st.session_state.df, pd.DataFrame):
+
+            df = st.session_state.df
+            selected_users = st.session_state.selected_users
+
+            st.markdown("For any criteria, you can see how user(s) place the cursor.")
+
+            selected_crit = st.selectbox("Select a criteria:", CRITERIA)
+
+            fig = go.Figure()
+
+            if len(selected_users):
+
+                for user in selected_users:
+                    df_user = df[df["public_username"] == user]
+                    fig.add_trace(go.Histogram(x=df_user[selected_crit], name=user, nbinsx=21))
+
+            else:
+                fig.add_trace(go.Histogram(x=df[selected_crit], name="all users", nbinsx=21))
+
+            fig.update_layout(barmode="overlay")
+            fig.update_traces(opacity=0.7)
+            st.plotly_chart(fig)
+
+        else:
+            st.warning(MSG_NO_DATA)
```

Comment:
Instead of indenting long if/else blocks in these functions, using an early return may be more readable:

```py
if not isinstance(st.session_state.df, pd.DataFrame):
  st.warning(MSG_NO_DATA)
  return

# Process data here...
```

Reply:
I confirm that the fail-fast / return early pattern in this kind of configuration greatly helps to understand what are the "requirements" of the function,  and gives the possibility to have the expected "positive" return at the end. 

There are several other functions in this file that could benefit of this little refactor.

Reply:
Yes, thanks for the suggestion. I couldn't use it before but now that it is break down into several functions it is better like this.